src/lib/y2users/config.rb

```diff
@@ -57,22 +56,90 @@ def system(reader: nil, force_read: false)
     end
 
     attr_reader :name
-    attr_accessor :users
-    attr_accessor :groups
 
-    def initialize(name, users: [], groups: [])
+    def initialize(name)
       @name = name
-      @users = users
-      @groups = groups
+
+      @users_manager = ElementManager.new(config)
+      @groups_manager = ElementManager.new(config)
+
       self.class.register(self)
     end
 
+    def users
+      users_manager.elements.dup.freeze
+    end
+
+    def groups
+      groups_manager.elements.dup.freeze
+    end
+
+    def attach(elements*)
+      elements.each { |e| attach_element(e) }
+    end
+
     def clone_as(name)
       config = self.class.new(name)
-      config.users = users.map { |u| u.clone_to(config) }
-      config.groups = groups.map { |g| g.clone_to(config) }
+
+      elements = users + groups
+      elements.each { |e| config.clone_element(e)  }
 
       config
     end
+
+  protected
+
+    def clone_element(element)
+      cloned = element.clone
+      cloned.assign_internal_id(element.id)
+
+      attach(cloned)
+    end
+
+  private
+
+    attr_reader :users_manager
+
+    attr_reader :groups_manager
+
+    def attach_element(element)
+      case element
+      when User
+        users_manager.attach(element)
+      when Group
+        groups_manager.attach(element)
+      else
+        raise "Wrong parameter: given #{element.class} but Y2Users::User or Y2Users::Group expected"
+      end
+    end
+
+    class ElementManager
+      attr_reader :elements
+
+      def initialize(config)
+        @config = config
+        @elements = []
+      end
+
+      # TODO: check whether the user/group is already attached
+      # TODO: check whether the user/group already exists ?
+      def attach(element)
+        element.assign_config(config)
+        element.assign_internal_id(next_id) if element.id.nil?
+
+        @elements << element
+      end
+
+    private
+
+      def next_id
+        @next_id ||= max_id(elements)
```

Comment:
I worry this is not much robust. E.g. consider that you have staging with newly attached user1 and then you have some dialog config that is cloned from system and it adds user2, then both get same id. I think we need global register of ids, so we ensure that it cannot accidentally be same.

Reply:
Right now, cloning always generates a new config:

~~~
config2 = config1.clone_as("staging")
~~~

and you should only compare configs that were generated by cloning. Otherwise, ids from a config would collide with ids from another (not cloned) config.

Reply:
I don't see a problem with that. If I got it correctly, the situation described by @jreidinger  is this:

- You start with a config called `system` with users read from the system.
- Then you clone it into another called `staging` and you add user1 there.
- Then you clone `system` again into another config called `another` and add user2 there.

So far, so good. What's the problem in having the same id in user1 and user2? I don't see anything wrong with that.

In the end you will do something like `System::Writer.new.write(staging, initial: system)` or maybe `System::Writer.new.write(another, initial: system)`. But for sure you will not do something like `System::Writer.new.write(another, initial: staging)` because it simply makes no sense.

Reply:
well, my concern is that I expect that id is unique. So I can trust that if I see in logs that I have user with id X, it will be that user in whole logs. Now I need to find in logs ID and to which config it belongs and if we teoretically chain configs e.g. in wizard with config_step1, config_step2... and having also staging that is there for abort, then having mixture of ids is kind of critical for me. I just said that if we simply have global counter that just increase, we can be sure that id is always unique and probably I would also prefer to have users and groups unique id, so it is will be again easier to grep in logs.

Reply:
I see @jreidinger's point about potentially confusing ids, although that has never been a problem so far in storage-ng, for example.

Anyway, I just wanted to point one thing. Chaining steps will still not introduce collision. config_step1 will be based on staging and config_step2 will be based on config_step1. So no collision there.

Reply:
I believe the current algorithm has a problem. Steps to reproduce:

- Start with an empty config.
- Create a first user, `@next_id` is initialized to 0 and then increased by 1.
- Create two more, `@next_id` is increased so it now contains a 3. The user ids are 1, 2 and 3.
- Clone that config into another one called `staging`.
- Detach the user with id 3 from `staging`
- Create a new user in `staging`. `@next_id` is initialized to 2 and then increased by 1.
- Result: `staging` contains a user with id 3 that is not the same than the corresponding one in the initial config.

Reply:
@ancorgs I think it is not reproducable as next_id is initialize once in staging to 3 and won't change later. But it can happen in that staging, step1 and step2. step1 detach user, step2 attach user and now staging and step2 contain user with same id.